saleor/plugins/openid_connect/utils.py

```diff
@@ -0,0 +1,219 @@
+    redirect_url = build_absolute_uri(f"/plugins/{plugin_id}/callback")
+
+    return prepare_url(urlencode(params), redirect_url)  # type: ignore
+
+
+def fetch_jwks(jwks_url) -> Optional[dict]:
+    response = None
+    try:
+        response = requests.get(jwks_url)
+        jwks = response.json()
+    except requests.exceptions.RequestException:
+        logger.exception("Unable to fetch jwks from %s", jwks_url)
+        raise AuthenticationError("Unable to finalize the authentication process.")
+    except json.JSONDecodeError:
+        content = response.content if response else "Unable to find the response"
+        logger.exception(
+            "Unable to decode the response from auth service with jwks. "
+            "Response: %s",
+            content,
+        )
+        raise AuthenticationError("Unable to finalize the authentication process.")
+    keys = jwks.get("keys", [])
+    if not keys:
+        logger.warning("List of JWKS keys is empty")
+    return keys
+
+
+def get_jwks_keys_from_cache_or_fetch(jwks_url: str) -> dict:
+    jwks_keys = cache.get(JWKS_KEY)
+    if jwks_keys is None:
+        jwks_keys = fetch_jwks(jwks_url)
+        cache.set(JWKS_KEY, jwks_keys, JWKS_CACHE_TIME)
+    return jwks_keys
+
+
+def create_jwt_token(id_payload: CodeIDToken, user: User, access_token: str,) -> str:
+    additional_payload = {
+        "exp": id_payload["exp"],
+        "oauth_access_key": access_token,
+    }
+    jwt_payload = jwt_user_payload(
+        user,
+        JWT_ACCESS_TYPE,
+        exp_delta=None,  # we pass exp from auth service, in additional_payload
+        additional_payload=additional_payload,
+    )
+    return jwt_encode(jwt_payload)
+
+
+def create_jwt_refresh_token(user: User, refresh_token: str, csrf: str):
+    additional_payload = {"oauth_refresh_token": refresh_token, "csrf_token": csrf}
+    jwt_payload = jwt_user_payload(
+        user,
+        JWT_REFRESH_TYPE,
+        # oauth_refresh_token has own expiration time. No need to duplicate it here
+        exp_delta=None,
+        additional_payload=additional_payload,
+    )
+    return jwt_encode(jwt_payload)
+
+
+def get_parsed_id_token(token_data, jwks_url) -> CodeIDToken:
+    id_token = token_data["id_token"]
+    keys = get_jwks_keys_from_cache_or_fetch(jwks_url)
+    try:
+        claims = jwt.decode(id_token, keys, claims_cls=CodeIDToken)
+        claims.validate()
+    except DecodeError:
+        raise AuthenticationError("Unable to decode provided token")
+    except JoseError:
+        raise AuthenticationError("Token validation failed")
+    return claims
+
+
+def get_or_create_user_from_token(claims: CodeIDToken) -> User:
+    user = User.objects.filter(email=claims["email"]).first()
+    if not user:
+        user = User.objects.create(
+            is_active=True,
+            email=claims["email"],
+            first_name=claims.get("given_name", ""),
+            last_name=claims.get("family_name", ""),
+        )
```

Comment:
```suggestion
    user, _ = User.objects.get_or_create(
        email=claims["email"],
        defaults = {
            "is_active"=True,
            "first_name"=claims.get("given_name", ""),
            'last_name"=claims.get("family_name", ""),
        }
    )
```

Reply:
It is not equal. See the comment from line: 140.  It filters over an email to make sure that the user doesn't exist. If we don't find an email we will create the new user. But if the user already exists we will check if the user is_active

Reply:
@fowczarek

Reply:
@korycins That's not safe, it's possible for another request to create the user right after your first SELECT checks that it does not exist. In that case, `create` will raise an exception. That's what `get_or_create` explicitly guards against.

Reply:
@patrys Yeap, thanks. It is already changed